tools/docker: Dockerfile for prplwrt RAX40 build

VladyslavTupikin opened the pull request:
Currently, we do not have dockerfile for building prplwrt for RAX40.
Without dockerfile developer must find and install all depednencies
in own environment. Eeach developer has own non-standard environment.
Which can cause hard-to-debug hard-to-reproduce build runtime errors.

Create new dockerfile and entrypoint script to prepare clear and unified
build environment for all developers.

Commit:
tools/docker: Dockerfile for prplwrt RAX40 build

Currently, we do not have dockerfile for building prplwrt for RAX40.
Without dockerfile developer must find and install all depednencies
in own environment. Eeach developer has own non-standard environment.
Which can cause hard-to-debug hard-to-reproduce build runtime errors.

Create new dockerfile and entrypoint script to prepare clear and unified
build environment for all developers.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
tools/docker: Script to start docker container

Script helps start docker container with user permissions
and correct work directory.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

tomereli commented on tools/docker/builder/openwrt/toolchains/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM ubuntu:16.04
```
Why specifically 16.04?

tomereli commented on tools/docker/builder/openwrt/toolchains/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM ubuntu:16.04
+
+RUN dpkg --add-architecture i386
```
Can you add a comment for this command? What does it do and what is it needed for?

VladyslavTupikin replied:
This command adds ability to install packages for i386 architecture.

tomereli commented on tools/docker/builder/openwrt/toolchains/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+RUN dpkg --add-architecture i386
+
+RUN apt-get -y update && apt-get install --no-install-recommends -y sudo git-core \
```
1. We prefer to split it to single package per line.
2. No need for sudo - you are root by default in docker
3. Are you sure we need all the tools below?

VladyslavTupikin replied:
1. I will fix it according to commentary
2. I will fix it according to commentary
3. Yes, I am sure, this Dockerfile we use for building OpenWRT.

tomereli commented on tools/docker/builder/openwrt/toolchains/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+perl automake autoconf fakeroot build-essential crash makedumpfile \
+kernel-wedge git-core libncurses5 libncurses5-dev libelf-dev asciidoc binutils-dev \
+bison imagemagick openssl cmake git-core expect xutils-dev genext2fs m4 qemu \
+libqt4-dev qt4-dev-tools python-setuptools \
```
I don't see why we need QT - care to explain?

MuteSpirit replied:
The most usable commands for package or kernel options selection are `make defconfig`, `make menuconfig` and so on which can work in console, but maybe some developers use [xconfig](https://en.wikipedia.org/wiki/Xconfig) for the same purposes. That tool requires Qt

tomereli replied:
IMO the whole interactive option should be removed from the `entrypoint.sh` - developers should prepare everything in the build server natively, and just use a docker container to run `make` - in which case we should limit the tools we install only to what is needed for building prplWrt.
See `tools/docker/builder/rdkb/toolchain/bitbake.sh` for how we do this for building RDKB - I don't think that prplWrt should be any different.

Actually, the only reason we did that for RDKB since it requires Ubuntu 16.04 for some reason and our build server is Ubuntu18.04 LTS which is too new. There is no such requirement for prplWrt for all I know - which raises the question, why do we need this PR in the first place?

Provided we do, I would expect this Dockerfile to be part of prplWrt and not prplMesh, but since for convenience we have this for RDKB then OK, I guess.

If we do need this, then I expect that the following to be done **outside** the container:
```
git clone https://git.prpl.dev/prplmesh/prplwrt.git
cd prplwrt
./scripts/feeds setup "src-git,packages,https://git.openwrt.org/feed/packages.git;openwrt-19.07" "src-git,intel_feed,https://git.prpl.dev/prplmesh/feed-intel.git;intel-19.07" "src-git,iwlwav,https://git.prpl.dev/prplmesh/iwlwav.git;iwlwav-19.07" "src-git,routing,https://github.com/openwrt-routing/packages.git"
./scripts/feeds update 
./scripts/feeds install -a -p intel_feed
./scripts/feeds install -a -p iwlwav
./scripts/feeds install intel_mips
./scripts/feeds install zmq
./scripts/gen_config.sh rax40
```
Not sure about the `gen_config.sh` part though - need to check if it depends on host tools.
Then, like `bitbake.sh`, add a build `tools/docker/builder/prplwrt/make.sh` which will run make from within a container while mapping the prplWrt clone folder, and the home folder if necessary, to the container.

tomereli commented on tools/docker/builder/openwrt/toolchains/entrypoint.sh:
```diff
@@ -0,0 +1,53 @@
+# Note: We don't create a user's home directory because it is already created by
+# mounting a host's user .ssh directory in 'docker run' command.
+groupadd -g ${GROUP_ID} ${GROUP_NAME}
+useradd -g ${GROUP_ID} -s "/bin/bash" -u ${USER_ID} ${USER_NAME}
```
Please refer to `toolchains/rdkb/Dockerfile` - I did this differently there also without creating home directory since I map the current user home directory into the container.

tomereli reviewed (commented):
I'm not sure how this should be used - I expect to see some script, which will be used to build the toolchain and call `docker container run` with the $HOME, $USER, $USER_ID, $GROUP_ID and $OPENWRT_ROOT as environment variables like we do in `tools/docker/builder/rdkb/toolchain/bitbake.sh`.
In general I prefer that this will be as close as possible to how the rdkb builder is implemented.

Commit:
tools/docker: Add instruction with usage

Step-by-step instruction for building docker image
and usage docker container.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

tomereli commented:
Looking really good so far - only missing an update to the readme, but I think you can go ahead and mark this as ready for review.

rmelotte reviewed (changes requested):
I didn't make specific review comments because not sure we want to achieve here.
Is it about being able to compile a development version of prplWrt (meaning with changes not part of any public repo) or only to be able to compile prplWrt locally?

In the former case, we could adapt the existing Dockerfile to:
- bind-mount the prplwrt repository instead of cloning it
- maybe add a possibility to do the build at runtime instead of when building the docker image (makes it easier to debug build failures)

In the latter case, it's already covered by https://github.com/prplfoundation/prplMesh/pull/902

In any case, I don't think we should add another Dockerfile, so marking as changes requested.

VladyslavTupikin commented:
@rmelotte 

Goal of this PR is to commit a Dockerfile which may be used by developers to locally build prplWRT for RAX40. Using one common Docker image to build prplWRT images for RAX40 should simplify build procedure for everyone (no need to look for dependencies and install them to local workstation) and also unifies build environment across all developers. Latter one help to avoid some per-station-specific build issues (wrong version of dependencies installed, etc.)

MuteSpirit commented on tools/docker/builder/openwrt/toolchains/entrypoint.sh:
```diff
@@ -0,0 +1,53 @@
+
+# Check if the command argument is a service command
+if [ "${cmd}" = "show-launch-script" ] ; then
+    cat /usr/local/bin/openwrt-env
```
I don't see any mention of `openwrt-env` script in your Dockerfile.
Did you mean `prplwrt-env`?
If yes, then please make it a part of docker image

mariomaz commented on tools/docker/builder/openwrt/toolchains/README.txt:
```diff
@@ -0,0 +1,42 @@
+
+4. Put prplwrt-env in /usr/bin/
+
+    mv prplwrt-env /usr/bin/
```
You need root access to move files to /usr/bin. I cannot do that in my box

mariomaz commented on tools/docker/builder/openwrt/toolchains/entrypoint.sh:
```diff
@@ -0,0 +1,53 @@
+./scripts/feeds install -a -p luci
+./scripts/feeds install intel_mips
+./scripts/feeds install zmq
+./scripts/gen_config.sh rax40 luci
```
I think this instructions are required to configure the build environment and thus should be executed in the first execution only.

tomereli replied:
I think these should be done natively in the build server if possible

mariomaz commented on tools/docker/builder/openwrt/toolchains/README.txt:
```diff
@@ -0,0 +1,42 @@
+
+3. After entering in docker container shell start building
+
+     make world -j4 V=99
```
Depends on the number of cores you have

tomereli replied:
Better to change to `make world -j$(nproc) V=s`

mariomaz reviewed (changes requested):
Other build scripts do not use docker interactive mode. IMO both possibilities should exist.
This docker image could also be adapted to build prplMesh for RAX40 (once you have already built prplWrt)

tomereli reviewed (changes requested):
I don't see the benefit of the interactive option, can you explain how you use it and why it is better than the approach taken in RDKB builder?

ghost commented:
Closing this PR due to replication of existing functionality
